mds/provider_poller/poller.py

```diff
@@ -189,8 +212,120 @@ def _poll_status_changes(self):
+                (f"\tLast event_time: {str(self.provider.last_event_time_polled)}.")
+                if not self.cursor
+                else f"\tCursor value: {self.from_cursor}."
+            )
+        )
+
+        params = {}
+
+        # Resume from the last "event_time" value
+        last_event_time_polled = self.provider.last_event_time_polled
+        if not last_event_time_polled:
+            last_event_time_polled = timezone.now() - datetime.timedelta(
+                getattr(settings, "PROVIDER_POLLER_LIMIT", 90)
+            )
+
+        # But we now apply a "lag" before actually polling,
+        # leaving time for the provider to collect data from its devices
+        polling_lag = self.provider.api_configuration.get("provider_polling_lag")
+        if polling_lag:
+            polling_lag = parse_duration(polling_lag)
+            if (timezone.now() - last_event_time_polled) < polling_lag:
+                logger.info("Still under the polling lag, back to sleep.")
+                return
+
+        # The MDS 0.4 Provider API got two endpoints:
+        # - /events for the real-time or so data (formerly /status_changes)
+        #   but limited to two weeks of history
+        # - /status_changes for all the data except the current hour
+        # If we're catching up far in time, begin by polling /status_changes
+        realtime_threshold = parse_duration(  # Default is 9 days
+            self.provider.api_configuration.get("realtime_threshold", "P9D")
+        )
+        if (timezone.now() - last_event_time_polled) > realtime_threshold:
+            # We have to query the archived status changes with another format
+            logger.info("last_event_time_polled is too old, asking archives")
+            params["event_time"] = last_event_time_polled.isoformat()[
+                : len("YYYY-MM-DDTHH")
+            ]
+        else:
+            # Query the real-time endpoint as usual
+            params["start_time"] = utils.to_mds_timestamp(
+                self.from_cursor or last_event_time_polled
+            )
+
+        # Provider-specific params to optimise polling
+        try:
+            params.update(self.provider.api_configuration["status_changes_params"])
+        except KeyError:
+            pass
+
+        endpoint = "events"  # The new name for the real-time events endpoint
+        if "event_time" in params:
+            # We asked the archived status changes instead
+            endpoint = "status_changes"
+        next_url = urllib.parse.urljoin(self.provider.base_api_url, endpoint)
+        if self.provider.api_configuration.get("trailing_slash"):
+            next_url += "/"
+
+        if params:
+            next_url = "%s?%s" % (next_url, urllib.parse.urlencode(params))
+
+        # Pagination
+        while next_url:
+            body = self._get_body(next_url, api_version)
+            # Translate older versions of data
+            translated_data = translate_data(body["data"], body["version"])
```

Comment:
I don't see what's that for since we now use separate functions for each version

Reply:
Only for fetching the status changes, the inner loop will only read a body the latest format.